x-pack/plugins/reporting/server/lib/content_stream.ts

```diff
@@ -249,8 +251,52 @@ export class ContentStream extends Duplex {
     });
   }
 
-  private async flush(size = this.buffer.byteLength) {
-    const chunk = this.buffer.slice(0, size);
+  private async flush(size = this.bytesBuffered) {
+    const buffersToFlush: Buffer[] = [];
+    let partiallyFlushedBuffer: undefined | Buffer;
+    let bytesToFlush = 0;
+
+    /*
+     Loop over each buffer, keeping track of how many bytes we have added
+     to the array of buffers to be flushed. The array of buffers to be flushed
+     contains buffers by reference, not copies. This avoids putting pressure on
+     the CPU for copying buffers or for gc activity. Please profile performance
+     with a large byte configuration and a large number of records (900k+)
+     before changing this code. Config used at time of writing:
+
+      xpack.reporting:
+        csv.maxSizeBytes: 500000000
+        csv.scroll.size: 1000
+
+     At the moment this can put memory pressure on Kibana. Up to 1,1 GB in a dev
+     build. It is not recommended to have overly large max size bytes but we
+     need this code to be as performant as possible.
+    */
+    for (const buffer of this.buffers) {
+      const remainder = size - bytesToFlush;
+      if (remainder < 0) {
+        break;
+      }
+      const chunkedBuffer = remainder > 0 ? buffer.slice(0, remainder) : buffer;
+
+      buffersToFlush.push(chunkedBuffer);
+      bytesToFlush += chunkedBuffer.byteLength;
+
+      // Did we add a partial buffer? Cache the rest and stop looping.
+      if (buffer.byteLength > remainder) {
+        partiallyFlushedBuffer = buffer.slice(remainder);
+        break;
+      }
+    }
```

Comment:
Looking at the logic inside the loop, I would suggest treating the `buffers` as a queue, and using `shift`/`unshift` instead of recreating the array.
```suggestion
    while (this.buffers.length) {
      const remainder = size - bytesToFlush;
      if (remainder < 0) {
        break;
      }

      const buffer = this.buffers.shift();
      const chunkedBuffer = remainder >= buffer.bytesLength ? buffer : buffer.slice(0, remainder);
      buffersToFlush.push(chunkedBuffer);
      bytesToFlush += chunkedBuffer.byteLength;
      
       if (buffer.byteLength > remainder) {
        this.buffers.unshift(buffer.slice(remainder));
      }
    }
```

With this approach, you don't need to recreate the array later, and the code seems more straightforward.

Reply:
This looks great! The only reason I think we should avoid mutating `this.buffers` here is to avoid making changes before we have actually flushed anything. WDYT about doing:

```ts
const buffers = this.buffers.slice(0);
...

this.buffers = buffers;
```

Reply:
It is a synchronous code. There is no need for that.

Reply:
OK, I was thinking of the calls further down to write to ES. But I see we don't actually recover from a failure to write (no try catch continuation logic). So if the network call to ES fails at some random point the entire export will fail.